server/src/internalClusterTest/java/org/elasticsearch/health/UpdateHealthInfoCacheIT.java

```diff
@@ -0,0 +1,116 @@
+            throw new RuntimeException("Failed to close internal cluster: " + e.getMessage(), e);
+        }
+    }
+
+    public void testNodeLeavingCluster() throws Exception {
+        try (InternalTestCluster internalCluster = internalCluster()) {
+            ClusterState state = internalCluster.client().admin().cluster().prepareState().clear().setNodes(true).get().getState();
+            Collection<DiscoveryNode> nodes = state.getNodes().getNodes().values();
+            DiscoveryNode healthNode = waitAndGetHealthNode(internalCluster);
+            DiscoveryNode nodeToLeave = nodes.stream()
+                .filter(node -> node.getId().equals(healthNode.getId()) == false)
+                .findAny()
+                .orElseThrow();
+            internalCluster.stopNode(nodeToLeave.getName());
+            assertBusy(() -> {
+                Map<String, DiskHealthInfo> healthInfoCache = internalCluster.getInstance(HealthInfoCache.class, healthNode.getName())
+                    .getDiskHealthInfo();
+                assertThat(healthInfoCache.size(), equalTo(nodes.size() - 1));
+                for (DiscoveryNode node : nodes) {
+                    if (node.getId().equals(nodeToLeave.getId())) {
+                        assertThat(healthInfoCache.containsKey(node.getId()), equalTo(false));
+                    } else {
+                        assertThat(healthInfoCache.get(node.getId()), equalTo(GREEN));
+                    }
+                }
+            });
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to close internal cluster: " + e.getMessage(), e);
+        }
+    }
+
+    public void testHealthNodeFailOver() throws Exception {
+        try (InternalTestCluster internalCluster = internalCluster()) {
+            ClusterState state = internalCluster.client().admin().cluster().prepareState().clear().setNodes(true).get().getState();
+            String[] nodeIds = state.getNodes().getNodes().keySet().toArray(new String[0]);
+            DiscoveryNode healthNodeToBeShutDown = waitAndGetHealthNode(internalCluster);
+            internalCluster.restartNode(healthNodeToBeShutDown.getName());
+            DiscoveryNode newHealthNode = waitAndGetHealthNode(internalCluster);
+            assertBusy(() -> {
+                Map<String, DiskHealthInfo> healthInfoCache = internalCluster.getInstance(HealthInfoCache.class, newHealthNode.getName())
+                    .getDiskHealthInfo();
+                assertThat(healthInfoCache.size(), equalTo(nodeIds.length));
+                for (String nodeId : nodeIds) {
+                    assertThat(healthInfoCache.get(nodeId), equalTo(GREEN));
+                }
+            });
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to close internal cluster: " + e.getMessage(), e);
+        }
+    }
+
+    private static DiscoveryNode waitAndGetHealthNode(InternalTestCluster internalCluster) throws InterruptedException {
```

Comment:
Shall we mark this with @Nullable? And the usages can possibly `assertNonNull` before using the health node?

Reply:
Hm, yes, it is cleaner this way.